File: pynes/tests/unit/fceux_runner.py
import socket
import subprocess
import time
import tempfile
import shutil

from os.path import abspath, dirname, join
from pynes.cart import Cart
from neslib.library import lib
from neslib.font import font_chr

HEADER_SIZE = 16
PRG_SIZE = 16384

SCRIPT_DIR = abspath(dirname(__file__))


class MemoryProxy:
    """Helper class to allow bracket and slice access to CPU RAM."""

    def __init__(self, runner):
        self.runner = runner

    def __getitem__(self, key):
        if isinstance(key, slice):
            start = key.start or 0
            stop = key.stop or 0x800
            step = key.step or 1
            return [
                self.runner._read_cpu_byte(addr)
                for addr in range(start, stop, step)
            ]
        return self.runner._read_cpu_byte(key)


class VramProxy:
    """Helper class to allow bracket and slice access to PPU VRAM."""

    def __init__(self, runner):
        self.runner = runner

    def __getitem__(self, key):
        if isinstance(key, slice):
            start = key.start or 0
            stop = key.stop or 0x4000
            step = key.step or 1
            return [
                self.runner._read_ppu_vram_byte(addr)
                for addr in range(start, stop, step)
            ]
        return self.runner._read_ppu_vram_byte(key)


class FCEUXRunner:
    def __init__(
        self, source, host='127.0.0.1', port=8888, fceux_path='fceux'
    ):
        self.host = host
        self.port = port
        self.fceux_path = fceux_path
        self.process = None
        self.conn = None

        # 1. Compile the source artifact identically to your original architecture
        cart = Cart(libraries=[lib], chr_banks=1, chr_data=font_chr())
        self.rom = cart.to_nes(source)

        prg = self.rom[HEADER_SIZE : HEADER_SIZE + PRG_SIZE]

        # Extract vectors to coordinate manual NMI triggers over the socket
        self.nmi_vector = prg[0x3FFA] | (prg[0x3FFB] << 8)

        # Wire structural proxy layers to match original headless syntax expectations
        self.cpu = type("ProxyCPU", (), {"memory": MemoryProxy(self)})()
        self.ppu = type("ProxyPPU", (), {"vram": VramProxy(self)})()

    def start_fceux(self, lua_script_path="fceux_runner.lua"):
        # with tempfile.TemporaryDirectory() as tmp:
        #     rom_path = join(tmp, 'test.nes')
        #     with open(rom_path, 'wb') as f:
        #         f.write(self.rom)
        rom_path = join('/tmp', 'test.nes')
        with open(rom_path, 'wb') as f:
            f.write(self.rom)
        """Launches FCEUX as a server and connects the Python client socket."""
        abs_lua_path = join(SCRIPT_DIR, lua_script_path)
        # Command configurations to invoke FCEUX headlessly with the server script
        cmd = [
            shutil.which('fceux'),
            # "--sound", "0",            # Disable audio processing
            # "--frameskip", "0",        # Enforce frame-by-frame lock step
            "--loadlua",
            abs_lua_path,  # Load the TCP socket server script
            # "--winsound", "0"
            rom_path,
        ]
        print(rom_path)
        self.stdout_log = open("fceux_stdout.log", "a", encoding="utf-8")
        self.stderr_log = open("fceux_stderr.log", "a", encoding="utf-8")

        # Start the background process running the server
        self.process = subprocess.Popen(
            cmd, stdout=self.stdout_log, stderr=self.stderr_log
        )

        # Allow FCEUX a moment to start and bind its listening port
        time.sleep(0.8)

        # Create a client socket and dial the emulator
        self.conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        retries = 5
        while retries > 0:
            try:
                self.conn.connect((self.host, self.port))
                break
            except ConnectionRefusedError:
                print(f"Connection refused, retries left: {retries}")
                time.sleep(0.5)
                retries -= 1

        if retries == 0:
            self.close_fceux()
            raise TimeoutError(
                "[RUNNER ERROR] Failed to connect to FCEUX server instance."
            )

        # Stream the compiled ROM data over to the server immediately after connecting
        # self.conn.sendall(f"LOAD:{len(self.rom)}\n".encode())
        # self.conn.sendall(self.rom)
        # print(self.conn.recv(4))
        # assert self.conn.recv(4) == b'ACK\n'

    def close_fceux(self):
        """Safely tears down the client socket connection and kills the FCEUX process."""
        if self.conn:
            try:
                self.conn.close()
            except:
                pass
            self.conn = None

        if self.process:
            self.process.terminate()
            try:
                self.process.wait(timeout=2.0)
            except subprocess.TimeoutExpired:
                self.process.kill()
            self.process = None

        if self.stdout_log:
            self.stdout_log.close()
        if self.stderr_log:
            self.stderr_log.close()

    def _read_cpu_byte(self, address):
        """Dispatches an atomic CPU RAM byte query to FCEUX."""
        self.conn.sendall(f"READ:{address}\n".encode())
        return int.from_bytes(self.conn.recv(1), byteorder='big')

    def _read_ppu_vram_byte(self, address):
        """Dispatches an atomic PPU VRAM byte query to FCEUX (read
        through the PPU registers at the frame boundary)."""
        self.conn.sendall(f"VRAM:{address}\n".encode())
        return self._recv_exact(1)[0]

    def read_pixels(self, x, y, width, height):
        """Reads an RGB block of the rendered screen: bytes in
        r, g, b order, row by row."""
        self.conn.sendall(f"PIXELS:{x},{y},{width},{height}\n".encode())
        return self._recv_exact(width * height * 3)

    def _recv_exact(self, size):
        data = b''
        while len(data) < size:
            chunk = self.conn.recv(size - len(data))
            if not chunk:
                raise ConnectionError('FCEUX closed the connection')
            data += chunk
        return data

    def press(self, buttons):
        """Applies the target button mask on the first controller layout."""
        self.conn.sendall(f"PADS:{buttons & 0xFF}\n".encode())
        assert self.conn.recv(4) == b'ACK\n'

    def release(self):
        """Releases all active keys on the pad layout."""
        self.press(0x00)

    def _run(self):
        """Tells FCEUX to execute instruction steps until a JMP-to-self loop parks."""
        self.conn.sendall(b"RUN_IDLE\n")
        status = self.conn.recv(10).decode().strip()
        if status == "RUNAWAY":
            raise RuntimeError('runaway execution (no idle loop reached)')
        if status != 'OK':
            raise RuntimeError('Not OK')

    def run_reset(self):
        """Runs the RESET vector execution block until the main loop parks."""
        self._run()

    def run_frames(self, frames):
        """Advances whole frames synchronously by force injecting NMIs."""
        for _ in range(frames):
            self.conn.sendall(b"NMI_ENABLED\n")
            if self.conn.recv(2).decode().strip() == "0":
                continue

            self.conn.sendall(f"NMI:{self.nmi_vector}\n".encode())
            status = self.conn.recv(10).decode().strip()
            if status == "RUNAWAY":
                raise RuntimeError('runaway execution (no idle loop reached)')

File: pynes/tests/unit/fceux_runner.lua
-- FCEUX Headless Runner Server Bridge with On-Screen Terminal Logger
local socket = require("socket")

local host = "127.0.0.1"
local port = 8888

local lines_log = {}
local MAX_LINES = 9
local LINE_HEIGHT = 10

local function screen_log(message)
    print(message)
    table.insert(lines_log, message)
    if #lines_log > MAX_LINES then
        table.remove(lines_log, 1)
    end
    emu.frameadvance()
end

local function draw_terminal_overlay()
    local start_x = 10
    local start_y = 15
    
    for i, line in ipairs(lines_log) do
        gui.text(start_x, start_y + ((i - 1) * LINE_HEIGHT), line, "green", "black")
    end
end

gui.register(draw_terminal_overlay)

-- Bind socket to port and start listening
local server = assert(socket.bind(host, port))
server:settimeout(5.0)

screen_log("[LUA SERVER] Awaiting Python client on port " .. port .. "...")
local client, err = server:accept()

if err or not client then
    screen_log("[LUA SERVER] Connection timeout or failure: " .. tostring(err))
    os.exit()
end

screen_log("[LUA SERVER] Connected to Python Test Harness.")
client:settimeout(nil) -- Switch to standard blocking mode for synchronized testing

local function get_pc()
    return memory.getregister("pc")
end

-- Simulates your Python _run() logic inside the emulator core
local function run_until_idle(max_frames)
    local last_pc = -1
    max_frames = 160
    for frame = 1, max_frames do
        emu.frameadvance()

        local current_pc = get_pc()
        
        if current_pc == last_pc then
            screen_log("[RUN_IDLE] CPU parked at PC=" .. string.format("0x%04X", current_pc) .. " after " .. frame .. " frames")
            return true
        end
        
        last_pc = current_pc
    end
    screen_log("[RUN_IDLE] Timeout after " .. max_frames .. " frames")
    return false
end


-- Main protocol server loop
while true do
    local line, err = client:receive("*l")
    if err then break end
    
    if line:sub(1, 5) == "LOAD:" then
        local size = tonumber(line:sub(6))
        local rom_data = client:receive(size)
        
        screen_log("[LOAD] Received new ROM binary block (" .. size .. " bytes)")
        
        local tmp_path = "tmp_runner_test.nes"
        local f = io.open(tmp_path, "wb")
        f:write(rom_data)
        f:close()
        
        emu.loadrom(tmp_path)
        client:send("ACK\n")
        screen_log("[LOAD] ROM flashed. Emulator Hard Reset.")
        
    elseif line == "RUN_IDLE" then
        screen_log("[EXEC] Running CPU until JMP-to-self idle...")
        local success = run_until_idle(100000)
        if success then
            client:send("OK\n")
            screen_log("[EXEC] CPU parked successfully.")
        else
            client:send("RUNAWAY\n")
            screen_log("[WARN] Runaway execution detected!")
        end
        
    elseif line:sub(1, 4) == "NMI:" then
        local nmi_vector = tonumber(line:sub(5))
        screen_log(string.format("[NMI] Forcing NMI trigger vector: $X", nmi_vector))
        
        local pc = get_pc()
        local status = memory.getregister("p")
        local s = memory.getregister("s")
        
        local pc_high = bit.band(bit.rshift(pc, 8), 0xFF)
        local pc_low  = bit.band(pc, 0xFF)

        memory.writebyte(0x0100 + s, pc_high)
        memory.writebyte(0x0100 + (s - 1), pc_low)
        memory.writebyte(0x0100 + (s - 2), status)
        memory.setregister("s", s - 3)
        memory.setregister("pc", nmi_vector)
        
        local success = run_until_idle(100000)
        if success then
            client:send("OK\n")
        else
            client:send("RUNAWAY\n")
        end
        
    elseif line:sub(1, 5) == "READ:" then
        local addr = tonumber(line:sub(6))
        local val = memory.readbyte(addr)
        client:send(string.char(val))

    elseif line:sub(1, 5) == "VRAM:" then
        local addr = tonumber(line:sub(6))
        local val = ppu.readbyte(addr)
        client:send(string.char(val))

    elseif line:sub(1, 7) == "PIXELS:" then
        local coords = {}
        for value in string.gmatch(line:sub(8), '[^,]+') do
            table.insert(coords, tonumber(value))
        end
        local x0, y0, w, h = coords[1], coords[2], coords[3], coords[4]
        local out = {}
        for y = y0, y0 + h - 1 do
            for x = x0, x0 + w - 1 do
                local r, g, b = emu.getscreenpixel(x, y, true)
                table.insert(out, string.char(r, g, b))
            end
        end
        client:send(table.concat(out))

    elseif line:sub(1, 5) == "PADS:" then
        local mask = tonumber(line:sub(6))
        screen_log(string.format("[INPUT] Applying controller mask: bitmask 0x%02X", mask))
        
        local buttons = {
            A      = bit.band(mask, 0x01) ~= 0,
            B      = bit.band(mask, 0x02) ~= 0,
            Select = bit.band(mask, 0x04) ~= 0,
            Start  = bit.band(mask, 0x08) ~= 0,
            Up     = bit.band(mask, 0x10) ~= 0,
            Down   = bit.band(mask, 0x20) ~= 0,
            Left   = bit.band(mask, 0x40) ~= 0,
            Right  = bit.band(mask, 0x80) ~= 0
        }
        joypad.set(1, buttons)
        client:send("ACK\n")
        
    elseif line == "NMI_ENABLED" then
        local ppuctrl = memory.readbyte(0x2000)
        if bit.band(ppuctrl, 0x80) ~= 0 then
            client:send("1\n")
        else
            client:send("0\n")
        end
    end
end
